src/components/tl/ucp/tl_ucp_sendrecv.h

```diff
@@ -225,20 +225,107 @@ static inline ucc_status_t ucc_tl_ucp_send_nz(void *buffer, size_t msglen,
+
+    for (; i < memh->num_tls; i++) {
+        if (strncmp(memh->tl_h[i].tl_name, "ucp", 3) == 0) {
+            *tl_index = i;
+            return UCC_OK;
+        }
+    }
+    return UCC_ERR_NOT_FOUND;
+}
+
+static inline ucc_status_t ucc_tl_ucp_get_memh(ucc_tl_ucp_team_t *team,
+                                               ucc_mem_map_mem_h  map_memh,
+                                               void             **ucp_memh)
+{
+    ucc_mem_map_memh_t     *memh     = map_memh;
+    int                     tl_index = 0;
+    ucc_tl_ucp_memh_data_t *tl_data;
+    ucc_status_t            status;
+
+    status = find_tl_index(memh, &tl_index);
+    if (status == UCC_ERR_NOT_FOUND) {
+        tl_error(UCC_TL_TEAM_LIB(team),
+                 "attempt to perform one-sided operation with malformed mem "
+                 "map handle");
+        return status;
+    }
+    tl_data   = (ucc_tl_ucp_memh_data_t *)memh->tl_h[tl_index].tl_data;
+    *ucp_memh = tl_data->rinfo.mem_h;
+    return UCC_OK;
+}
+
+static inline ucc_status_t ucc_tl_ucp_check_memh(ucp_ep_h *ep, ucc_rank_t me,
+                                                 ucc_rank_t peer, void *va,
+                                                 uint64_t *rva, ucp_rkey_h *rkey,
+                                                 int tl_index,
+                                                 ucc_mem_map_mem_h *dst_map_memh)
+{
+    ucc_mem_map_memh_t     **dst_memh   = (ucc_mem_map_memh_t **)dst_map_memh;
+    ucc_tl_ucp_memh_data_t *dst_tl_data = NULL;
+    uint64_t                base;
+    uint64_t                end;
+    ucs_status_t            ucs_status;
+    int                     i;
+    size_t                  offset;
+
+    base = (uint64_t)dst_memh[me]->address;
+    end  = base + dst_memh[me]->len;
+
+    if (!((uint64_t)va >= base && (uint64_t)va < end)) {
+        return UCC_ERR_NOT_FOUND;
+    }
+    *rva = (uint64_t)PTR_OFFSET(dst_memh[peer]->address,
+                                ((uint64_t)va - (uint64_t)dst_memh[me]->address));
+    dst_tl_data = (ucc_tl_ucp_memh_data_t *)dst_memh[peer]->tl_h[tl_index].tl_data;
+    if (NULL == dst_tl_data->rkey) {
+        offset = 0;
+        /* find pack location for tl */
+        for (i = 0; i < tl_index; i++) {
+            char *name          = PTR_OFFSET(dst_memh[peer]->pack_buffer, offset);
+            size_t *packed_size = PTR_OFFSET(dst_memh[peer]->pack_buffer, offset
+                                + UCC_MEM_MAP_TL_NAME_LEN);
+            if (strncmp(name, "ucp", 3) == 0) {
+                break;
+            }
+            offset += *packed_size;
+        }
+        ucs_status = ucp_ep_rkey_unpack(
+            *ep, PTR_OFFSET(dst_memh[peer]->pack_buffer, offset +
+                    sizeof(size_t) *
+                    (UCC_TL_UCP_MEMH_TL_HEADERS + UCC_TL_UCP_MEMH_TL_PACKED_HEADERS)),
+            &dst_tl_data->rkey);
+        if (UCS_OK != ucs_status) {
+            return ucs_status_to_ucc_status(ucs_status);
```

Comment:
do we need to free the keys from previous iterations with `ucp_rkey_destroy`?

Reply:
I don't think so. `dst_tl_data` is specific to a particular peer. We should only go down this code path once per peer and future calls should skip the `if (NULL == dst_tl_data->rkey)` statement. Eventually the rkeys should be destroyed in `ucc_mem_unmap`.